tech_soln/real_time_analytics.rst

```diff
@@ -0,0 +1,425 @@
+HLLs this query involves shipping lists of ip addresses from the workers to the master for
+it to deduplicate. That's both a lot of network traffic and a lot of computation. By using
+HLLs you can greatly improve query speed.
+
+First you must install the hll extension; `the github repo
+<https://github.com/aggregateknowledge/postgresql-hll>`_ has instructions. Next, you have
+to enable it:
+
+.. code-block:: sql
+
+  -- this part must be run on all workers
+  CREATE EXTENSION hll;
+
+  -- this part runs on the master
+  ALTER TABLE http_requests_1min ADD COLUMN distinct_ip_addresses (hll);
+
+When doing our rollups, we can now aggregate sessions into an hll column with queries
+like this:
+
+.. code-block:: sql
+
+  SELECT
+    zone_id, date_trunc('minute', ingest_time) as minute,
+    hll_add_agg(hll_hash_text(ip_address)) AS distinct_ip_addresses
+  WHERE minute = date_trunc('minute', now())
+  FROM http_request
+  GROUP BY zone_id, minute;
+
+Now dashboard queries are a little more complicated, you have to read out the distinct
+number of ip addresses by calling the ``hll_Cardinality`` function:
+
+.. code-block:: sql
+
+  SELECT
+    request_count, success_count, error_count, average_response_time_msec,
+    hll_cardinality(distinct_ip_addresses) AS distinct_ip_address_count
+  FROM http_request_1min
+  WHERE zone_id = 1 AND minute = date_trunc('minute', now());
+
+HLLs aren't just faster, they let you do things you couldn't previously. Say we did our
+rollups, but instead of using HLLs we saved the exact unique counts. This works fine, but
+you can't answer queries such as "how many distinct sessions were there during this
+one-week period in the past we've thrown away the raw data for?".
+
+With HLLs, this is easy. You'll first need to inform Citus about the ``hll_union_agg``
+aggregate function and its semantics. You do this by running the following:
+
+.. code-block:: sql
+
+  -- this should be run on the workers and master
+  CREATE AGGREGATE sum (hll)
+  (
+    sfunc = hll_union_agg,
+    stype = internal,
+  );
+
+Now, when you call SUM over a collection of HLLs, PostgreSQL will return the HLL for us.
+You can then compute distinct ip counts over a time period with the following query:
+
+.. code-block:: sql
+
+  -- working version of the above query
+  SELECT
+    hll_cardinality(SUM(distinct_ip_addresses))
+  FROM http_request_1min
+  WHERE ingest_time BETWEEN timestamp '06-01-2016' AND '06-28-2016';
+
+You can find more information on HLLs `in the project's GitHub repository
+<https://github.com/aggregateknowledge/postgresql-hll>`_.
+
+Where the HLL extension provides distinct counts, there are more extensions which do a
+similar thing (improve performance and storage requirements) for other kinds of queries,
+such as `count-min sketch <https://github.com/citusdata/cms_topn>`_ for top-n queries, and
+`HDR <https://github.com/citusdata/HDR>`_, for percentile queries.
```

Comment:
We aren't going to be releasing CMS topN and HDR in this release cycle. Could we remove it?